plugins/catalog-react/src/components/EntityAutocompletePicker/EntityAutocompletePicker.tsx

```diff
@@ -29,17 +29,22 @@ import {
 } from '../../hooks/useEntityListProvider';
 import { EntityFilter } from '../../types';
 
-type KeysMatchingCondition<T, V, K> = T extends V ? K : never;
-type KeysMatching<T, V> = {
+/** @public */
+export type KeysMatchingCondition<T, V, K> = T extends V ? K : never;
+
+/** @public */
+export type KeysMatching<T, V> = {
   [K in keyof T]-?: KeysMatchingCondition<T[K], V, K>;
 }[keyof T];
 
-type AllowedEntityFilters<T extends DefaultEntityFilters> = KeysMatching<
+/** @public */
+export type AllowedEntityFilters<T extends DefaultEntityFilters> = KeysMatching<
```

Comment:
I wonder if we can inline some of these types into one type rather than exporting 4 different types.

They're not used apart from in this file, so I'm thinking that although it might be a little hard to understand and read, I'd rather only export one type from the API surface, as it makes it much easier to maintain and grow over time.

Reply:
It took me quite a while to understand the Typescript black magic 😂 But I think it should work now.

Reply:
Following @benjdlambert I think we could also avoid exporting `AllowedEntitiesFilters` and `ConstructableFilter`:

```
export type EntityAutocompletePickerProps<
  T extends DefaultEntityFilters = DefaultEntityFilters,
  Name extends KeysMatching<
    T,
    EntityFilter & { values: string[] }
  > = KeysMatching<T, EntityFilter & { values: string[] }>,
> = {
  label: string;
  name: Name;
  path: string;
  showCounts?: boolean;
  Filter: {
    new (...args: unknown[]): EntityFilter;
  };
  InputProps?: TextFieldProps;
  initialSelectedOptions?: string[];
};

/** @public */
export function EntityAutocompletePicker<
  T extends DefaultEntityFilters = DefaultEntityFilters,
>(props: EntityAutocompletePickerProps<T>) {
```

this should allows us to completely delete `type AllowedEntityFilters` and `interface ConstructableFilter<T>`